File: WeightRecord/Controllers/ViewController.m
//
//  ViewController.m
//  WeightRecord
//
//  Created by Casey Yost on 2/1/17.
//  Copyright © 2017 Toasty. All rights reserved.
//

#import "AppDelegate.h"
#import "ViewController.h"
#import "WRCategory.h"
#import "WRExercise.h"

@implementation ViewController

- (instancetype)initWithNodes:(NSArray <WRNode *> *)nodes
{
    self = [super initWithNibName:nil bundle:nil];
    if(self)
    {
        _nodes = nodes;
    }
    return self;
}

// If we're getting this initializer, it means we've been initialized from the storyboard,
// so we're the root
- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if(self)
    {
        AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
        _nodes = appDelegate.rootNodes;
    }
    return self;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.tableView reloadData];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"CategoryCell"];
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"ExerciseCell"];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if(indexPath.row >= _nodes.count)
    {
        return;
    }
    
    WRNode *node = _nodes[indexPath.row];
    
    BOOL isCategory = [node isKindOfClass:[WRCategory class]];
    
    if(isCategory)
    {
        ViewController *vc = [[ViewController alloc] initWithNodes:[(WRCategory *)node subnodes]];
        [self.navigationController pushViewController:vc animated:YES];
    }
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return _nodes.count;
}

// Row display. Implementers should *always* try to reuse cells by setting each cell's reuseIdentifier and querying for available reusable cells with dequeueReusableCellWithIdentifier:
// Cell gets various attributes set automatically based on table (separators) and data source (accessory views, editing controls)

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if(indexPath.row >= _nodes.count)
    {
        return nil;
    }
    
    WRNode *node = _nodes[indexPath.row];
    BOOL isCategory = [node isKindOfClass:[WRCategory class]] ? YES : NO;
    NSString *identifier = isCategory ? @"CategoryCell" : @"ExerciseCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    
    if(cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:identifier];
    }
    
    if(isCategory)
    {
        WRCategory *category = (WRCategory *)node;
        cell.textLabel.text = category.name;
    }
    else
    {
        WRExercise *exercise = (WRExercise *)node;
        cell.textLabel.text = exercise.name;
    }
    
    return cell;
}

@end

File: WeightRecord/ViewControllers/ViewController.swift
//
//  ViewController.swift
//  WeightRecord
//
//  Created by Casey Yost on 9/8/18.
//  Copyright © 2018 Casey Yost. All rights reserved.
//

import UIKit

class ListViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    private var category: Category? = nil
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let tableView = UITableView(frame: .zero)
        tableView.delegate = self
        tableView.dataSource = self
        tableView.translatesAutoresizingMaskIntoConstraints = false
        
        self.view.addSubview(tableView)

        var constraints = NSLayoutConstraint.constraints(withVisualFormat: "|[tableView]|", options: [], metrics: nil, views: ["tableView": tableView])
        constraints += NSLayoutConstraint.constraints(withVisualFormat: "V:|[tableView]|", options: [], metrics: nil, views: ["tableView": tableView])
        
        NSLayoutConstraint.activate(constraints)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let category = category {
            return category.subCategories.count
        }
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var tableViewCell = tableView.dequeueReusableCell(withIdentifier: "Cell")
        if(tableViewCell == nil) {
            tableViewCell = ExerciseTableViewCell(Exercise())
        }
    
        return tableViewCell!
    }
}

